### ISSUE TITLE
Threading, Daemons, and Servers

### ISSUE THREAD

**Issue by Cabalist** [kizniche/mycodo_python#2]
I am going to try and keep my questions about threading here.  

First!  The Daemon claims to ```receive/execute commands from mycodo-client.py``` but it appears that mycodo_client has it's own rpyc_client that talks to the rpyc server.  How is the the daemon involved?  Is it?

**Comment by kizniche** [comment]
It used to be more integrated. It now only directly communicates to terminate the daemon. Otherwise, it merely executes the exposed_ functions from mycodo.py, which makes use of the globals.

**Comment by kizniche** [comment]
The daemon is instructed to terminate when `client_que = 'TerminateServer'`

which is set by:

```python
    def exposed_terminate_daemon(self, remoteCommand):
        global client_que
        client_que = 'TerminateServer'
        logging.info("[Client command] Shut down the daemon")
        mycodo_log.Concatenate_Logs()
        return 1
```

**Comment by Cabalist** [comment]
Gotcha!  This confirms what I saw yesterday.

**Comment by Cabalist** [comment]
Would it be safe to say mycodo (without the web part) is broken up like this:

An rpyc server that handles communication to and from the hardware

A daemon that runs periodically that calls functions on the rpyc server and writes out logs.

Interface functions that other programs (php server, CLI, etc) call that interact with the rpyc server.

**Comment by kizniche** [comment]
The rpyc server merely acts as a way to have mycodo_client (user executed) convey information to mycodo_server (root executed), separating the user from being able to execute any code as root.

The daemon essentially conducts sensor measurements and write logs. Other aspects like relay switching, PID control, sensor/relay conditionals, etc. are ancillary to that main functionality.

And the interface functions rely heavily on PHP DB modification and mycodo_client notifying mycodo_server of DB/PID/Relay changes.

**Comment by Cabalist** [comment]
Hmm... So the PHP code has access to the DB as well?

**Comment by kizniche** [comment]
Take a look at includes/restricted.php and if I listen carefully, I may be able to hear your jaw drop from here.

**Comment by kizniche** [comment]
...Maybe. I'm not sure how deep your disdain for PHP goes.

**Comment by Cabalist** [comment]
whoa.  Found the other half of everything.

**Comment by Cabalist** [comment]
I'm ... going to ignore that at the moment.  It's a little overwhelming.  

Does the PHP call any Python functions directly?  Or just through mycodo_client?

**Comment by kizniche** [comment]
All through mycodo_client

**Comment by kizniche** [comment]
Mostly with this `shell_exec("$mycodo_client --sqlreload -1");`

**Comment by Cabalist** [comment]
What keeps two processes from accessing a sensor at the same time?

**Comment by kizniche** [comment]
The sensor lock file

**Comment by Cabalist** [comment]
Does the daemon run as root?

**Comment by kizniche** [comment]
The sensor lock files. Although the sensor lock file system has room for improvement. Currently, only one of each type can use the lock. So one temperature sensor will wait until th eother temperature sensor has finished, even though they may not necessarily need to wait (if they are different sensors that use different routes to communicate with the devices).

**Comment by kizniche** [comment]
Yes, it is executed by Cron at bootup, as root.

**Comment by Cabalist** [comment]
And runs continuously?  i.e. ```while True```

**Comment by kizniche** [comment]
Yes, that loop is technically the daemon part.

**Comment by Cabalist** [comment]
Gotcha. I guess the next question is obvious:  What happens if it crashes?

**Comment by kizniche** [comment]
Shhhhh!

**Comment by kizniche** [comment]
I haven't gotten around to taking care of that. I've briefly looked at 3rd part software to handle that (http://linux.die.net/man/8/watchdog) or creating my own watch program.

**Comment by kizniche** [comment]
I've had the daemon run for months without crashing. I've had it run minutes and crash. Its stable when you're not changing global variables ;)

**Comment by kizniche** [comment]
There is at least one bug that stops the PID controller (major issue!) but it's elusive. I have a hunch it's linked to the use of globals in some way.

**Comment by Cabalist** [comment]
haha well.  No more globals when this is done.  

I am looking at reimplementing the Daemon using this:
https://pypi.python.org/pypi/python-daemon/
Documentation is kind of light but here is an example script:
http://www.gavinj.net/2012/06/building-python-daemon-process.html

**Comment by Cabalist** [comment]
Or possibly Twisted (https://twistedmatrix.com/trac/) to handle multi process communication

**Comment by Cabalist** [comment]
I need to think more about this.

**Comment by kizniche** [comment]
That's great. Mycodo quickly outgrew it's shoes and I just kept the old system instead of investing time in getting the foundation correct.

**Comment by Cabalist** [comment]
Does the php ever access the sensors?

**Comment by kizniche** [comment]
Only the log files. The daemon is the only part that communicates with sensors.

**Comment by Cabalist** [comment]
Does it use the lock files?

**Comment by Cabalist** [comment]
Is it ReadOnly with the Log files?

**Comment by kizniche** [comment]
The PHP only reads the log files.

**Comment by Cabalist** [comment]
And just to make sure the chain of control looks like this:

I want to change a relay:
UI -> PHP -> mycodo_client -> mycodo_server -> python_function -> hardware

I want to make a graph:
UI -> PHP -> mycodo_client -> mycodo_server -> python_function -> Log_files

I want to set a conditional for a Relay:
UI -> PHP -> DB

I want to view recent values of a sensor:
UI -> PHP -> Logs


Any other examples that you can provide?

**Comment by kizniche** [comment]
I would add the following for setting a conditional for a Relay:
UI -> PHP -> DB
UI -> PHP -> mycodo_client -> mycodo_server (notify of DB change)

It would be a lot easier if I could communicate the exact data that changed on the UI to the daemon (sanitized, of course). Ive had plans to figure this out, because it would make things easier and more efficient.

For one, I could more effectively manage the PID controllers because I would know what variables had changed and could pass only those to the PID thread. Currently, if I were to change the P, I, or D gains, I wouldn't have to restart the PID. I would have to reload all the variables because I have no system to determine if the new variable values are different from the previous, so I assume any or all could have changed.

I want to get the most current reading of a sensor:
mycodo_client -> mycodo_server -> python_function -> hardware
```
./mycodo-client.py --sensorht DHT22 1
```

From restricted.php:
UI -> PHP -> mycodo-wrapper -> init.d script -> mycodo-client -> mycodo_server
```php
exec("$install_path/cgi-bin/mycodo-wrapper restore $restore_directory >> /var/www/mycodo/log/restore.log &");
```
---
UI -> PHP -> mycodo-wrapper -> restore-mycodo.sh
```php
exec("$install_path/cgi-bin/mycodo-wrapper restart 2>&1 > /dev/null");
```

**Comment by Cabalist** [comment]
I think you can TRIGGER a relay with the mycodo_client but you can't set one in the DB, correct?

**Comment by Cabalist** [comment]
What if instead of one big Daemon we just spun off threads that read from sensors and logged the results?

We'd have a Controller class that would be able to pause, stop, and start the threads.  Adding a sensor would be as easy as passing in some values to the Controller and it would start a thread and a log file going for that sensor.

**Comment by kizniche** [comment]
> can't set one in the DB

Can you elaborate what you mean by this?

**Comment by kizniche** [comment]
> spun off threads that read from sensors and logged

This is what I was working toward with starting to modularize things. I knew I needed to get rid of globals to do this.

**Comment by Cabalist** [comment]
The UI populates values in the RelayConditional table for instance.  This is the only way to setup a conditional.  I don't believe mycodo_client (or anything else) sets up or changes the values in the RelayConditional table.

**Comment by kizniche** [comment]
Correct. Only the UI (and update script) modifies the DB.

**Comment by kizniche** [comment]
> Currently, if I were to change the P, I, or D gains, I wouldn't have to restart the PID thread.

Oops. That should have read "Currently, if I were to change the P, I, or D gains, I would have to restart the PID thread."

**Comment by Cabalist** [comment]
Cool!  Let's start moving toward a controller with threads that read sensors.

This would mean roughly:
- Creating a read_sensor() function that takes arguments to configure how the sensor is to be read.
- Creating a thread controller that will keep track of the threads and allow you to start new ones.

We are looking for these features:
- Ability to add arbitrary sensors.  Meaning the logic to read new sensors is decoupled from the read_sensor() function
- Ability to communicate to threads and for instance change PID gains

-----
You raise a good point actually.  mycodo_core doesn't write to the DB.  It is only reading settings and configuration data.  Fascinating.  
-----
I just keep saying these things over and over until they are firm in my head:

rpyc Server is started so that the mycodo_client has something to talk to.  The Server runs as root the Client runs as a user.  

Why does the server try and access the sensor?  Why doesn't it pop the last log entry off and return that?

**Comment by kizniche** [comment]
> Why does the server try and access the sensor? Why doesn't it pop the last log entry off and return that?

The PID controller operates more effectively the closer the input is to the present moment (unless you can find a way to see the future measurement), so I believe this necessitates the server performing the reading and passing it immediately to the PID controller to receive the output value to activate the relay, then it can send the value to be logged.

**Comment by Cabalist** [comment]
How much of a delay is acceptable?

**Comment by Cabalist** [comment]
If the daemon is running wouldn't a call from mycodo_client have to break a lock?  With a 60 second timeout?

**Comment by kizniche** [comment]
I wouldn't want more than a second from when the reading is acquired to when it's passed to the PID controller.

**Comment by kizniche** [comment]
> If the daemon is running wouldn't a call from mycodo_client have to break a lock?
Hm? What kind of call and what lock?

**Comment by Cabalist** [comment]
Earlier I asked:
>What keeps two processes from accessing a sensor at the same time?

You said:
>The sensor lock file

**Comment by kizniche** [comment]
Ah. mycodo_client will get stuck waiting until the lock is free, but it normally doesn't take more than a few seconds to get a reading and verify it (2 readings compared), so the lock will end quickly.

**Comment by kizniche** [comment]
The timeout is there in case for some reason the lock file remains but there's no process behind it, it will break the lock.

**Comment by Cabalist** [comment]
Does the daemon do the 2 readings compared thing?

If we stuck the readings into a DB the get_sensor_value from mycodo_client would be near instantaneous and always within a second (if the read_sensor() thread was running)

**Comment by kizniche** [comment]
We have:
```
Daemon
    read_t_sensor() <- does comparison
        read_t() <- reads actual sensor
```

The daemon takes periodic readings and logs them. If you read the logged value, its age could be a duration up to the duration of the period. Your initial question sounded like you wanted to just look to the log for the current reading. Because the log has aged data, it can't be used with the PID controllers, so they get a fresh reading. You could have mycodo_client read from the log, but I intended --sensor to be used to test sensors that are connected but don't necessarily have graphing/logging activated.

**Comment by Cabalist** [comment]
Gotcha.  

I am just playing around with different ways of doing this to make the abstraction layers a little more discrete and consistent.

It makes sense that the thread may be taking readings that are not at 1 second intervals.  

The daemon right now only holds the lock for how ever long it is taking the reading and when it tries to take a reading it must acquire the lock.  On top of that it is doing the readings of each sensor one after the other.  Meaning sensor2 has to wait for the lock to be broken and the reading to be stabilized on sensor1 before it can be read.  This *probably* leads to irregularities in the logs.

In my mind for the Controller-Thread model, the thread held the lock for the entire duration of it's life.  This would ensure that the sensor would always be ready when the thread woke up to do a reading.

Threads also ensure that sensor1 could be being read at the same time as sensor2.

**Comment by kizniche** [comment]
Yes, the queue for reading sensors in the daemon's while loop leads so sometimes significant irregularities.

I would have moved sensor reading to threads sooner if I understood threading more. I put it as a low priority however, because the PID controller was already so good, I saw the gains to moving to threading were minimal in most situations.

**Comment by Cabalist** [comment]
Understood.  Everyone I know has rough edges when it comes to multithreading.  It is effin hard.

Can you walk me through the way you use the PID Controllers?

**Comment by Cabalist** [comment]
We are getting into it today, huh?

**Comment by Cabalist** [comment]
I assume they are hooked up to heaters (or A/C), fans and maybe a humidifier?

**Comment by kizniche** [comment]
There is a `Measure Interval` for each PID controller, that determines the sensor measurement period. When a PID controller is turned on, a thread is started and the PID class initialized. The set point, P, I, and D values are set, then at every measurement, the specific condition (such as temperature) is input into the PID, and an output is returned that will be the duration the relay will be turned on for. The device will have to be chosen/tweaked to be able to modify the specific environmental condition in the direction you choose. For instance, a heater will raise temperature, cooler will lower temperature. The output duration will be specific to whether the condition needs to rise (turn heater relay on) or lower (turn cooler relay on).

**Comment by kizniche** [comment]
Since the PID output can be greater than the period, I built in the functionality of just continuing a relay on duration and logging the time that already elapsed, so the relay doesn't need to turn off just to turn back on, and I get the most accurate amount of time that relay was running before the next PID update provided a new output.

**Comment by Cabalist** [comment]
I see.  

I am going to simplify some more of this code and get the read_sensor stuff running without globals.  This means that this code may lose some features that are dependent on the globals as we refactor.  Primarily having to do with the daemon.  

I think I'll make a bunch of functions that take arguments and create output and once that is done I'll start working on a controller to make that multi-threaded.  It is good to know that that is the trajectory. 

I still think that putting a buffer in front of the sensors is the most reliable way to consistent readings from sensors accessed by multiple processes.  I think having the client access them directly as part of the general application is too fraught with peril. 

Cutting away bits and understanding it is showing me features (and issues) that were not apparent before.  Again I really appreciate the speed that you respond.

**Comment by Cabalist** [comment]
Haha.  Just got to the part in the daemon where you rolled your own signalling using globals.   This is remarkable.  I keep finding things in here that either show you are familiar with design patterns or very very clever.   TOO clever in places.  haha
